source/octod/media.d

```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+
+    Wraps media type string used for interaction with GitHub as defined
+    by https://developer.github.com/v3/media
+
+    Copyright: Copyright (c) 2016 Sociomantic Labs. All rights reserved
+
+    License: Boost Software License Version 1.0 (see LICENSE for details)
+
+*******************************************************************************/
+
+module octod.media;
+
+/**
+    Stores sub-parts of media type string as separate fields. Simplifies
+    branching based on media types and safeguards against some typos in the
+    string.
+ **/
+struct MediaType
+{
+    /// Media type used by default if none is specified by user of the library
+    enum Default = MediaType.init;
+
+    immutable
+    {
+        /**
+            Indicates if this media type is of "vnd.github" ones. If set to
+            `false`, this becomes simple `application/something` media type.
+         **/
+        bool   github_marker = true;
+        /**
+            API version to be used for the request. If empty, GitHub will
+            assume whatever default API version currently is.
+         **/
+        string ver           = "v3";
+        /**
+            Custom parameters, usually indicate kind of content to serve
+            (for example, rendered markdown or raw source)
+         **/
+        string param         = "";
+        /**
+            Format of response data, usually "json" or "xml"
+         **/
+        string format        = "json";
+    }
+
+    /**
+        Constructor for constructing most common media type
+        directly (as opposed to string parsing)
+
+        Params:
+            format = format of response data (for example "json")
+            param = any custom parameters (for example "raw")
+     **/
+    this ( string format, string param = "" )
+    {
+        this.param = param;
+        this.format = format;
+    }
+
+    /**
+        Constructor for arbitrary setting of all fields, private usage
+        only
+
+        Params:
+            format = format of response data (for example "json")
+            param = any custom parameters (for example "raw")
+     **/
+    private this ( bool github_marker, string ver, string param, string format )
+    {
+        this(format, param);
+        this.github_marker = github_marker;
+        this.ver = ver;
+    }
+
+    /**
+        Returns:
+            string representation that can be used as `Accept` HTTP header
+     **/
+    string toString ( )
+    {
+        import std.format;
+
+        if (!this.github_marker)
+            return std.format.format("application/%s", format);
+        else
+        {
+            return std.format.format(
+                "application/vnd.github%s%s%s",
+                this.ver.length ? "." ~ this.ver : "",
+                this.param.length ? "." ~ this.param : "",
+                this.format.length ? "+" ~ this.format : ""
```

Comment:
Well, we all know how this ends up... Is harmless until some user cares about allocations. Would it be too hard to keep an internal buffer for formatting? Or this is too old school and nobody will ever expect a D2 library not to allocate like mad? :)